docs/develop/guides/compact-javascript-runtime.mdx

```diff
@@ -0,0 +1,441 @@
+
+### **4\. Writing Unit Tests**
+
+Because the Compact runtime is just a CommonJS module, you can easily integrate it with **Vitest, Jest**, **Mocha**, or any testing framework you prefer.
+
+```typescript
+describe('Bulletin Board Contract', () => {  
+	it('should accept a new post', () => {  
+		const ctx = { originalState: { status: State.VACANT }, transactionContext: {} };
+
+    const result = contract.circuits.post(ctx, "Test message");
+
+    expect(result.result.status).toBe(State.OCCUPIED);  
+	});  
+});
+```
+
+This makes it easy to test your contract logic off-chain, with full control over inputs and without having to run a Midnight node or a proof server.
+
+The runtime thus serves as your **development playground**: a place to test, simulate, and reason about Compact contract logic before deploying it.
+
+In the next section, we’ll look at **why this design matters**, how the JavaScript runtime bridges human-readable code and zero-knowledge execution, and why it’s key to making privacy-preserving smart contracts developer-friendly.
+
+## **Why This Design Matters**
+
+In the previous section, we saw how you can import the generated runtime, instantiate your contract, and call its functions just like any normal JavaScript class.
+
+But you might be wondering — *why does Compact bother generating a JavaScript runtime at all?*
+
+Why not just compile directly to zero-knowledge circuits and leave it at that?
+
+The answer comes down to **developer experience**, **reproducibility**, and **trust**.
+
+### **1\. A Bridge Between ZK Circuits and Everyday Code**
+
+Zero-knowledge circuits are powerful, but they’re also complex and opaque, not something you can easily debug or test directly.
+
+The JavaScript runtime acts as a human-friendly bridge between the low-level proof system and the high-level contract logic.
+
+When you call:
+
+`const { result, proofData } = contract.circuits.post(context, "Hello world!");`
+
+You’re running exactly the same logic that the ZK circuit will later prove on-chain, but in a form that you can step through, log, and inspect in Node.js.
+
+This means you can **validate the behavior of your contract locally** before you ever need to touch the proof server or submit a transaction to the Midnight network.
+
+### **2\. Type Safety and Consistency Across Environments**
+
+Because the runtime uses Compact’s own type descriptors (`CompactTypeBoolean`, `CompactTypeBytes`, etc.), the data you pass into your JavaScript tests is encoded in the exact same way it will be on-chain.  That consistency eliminates a whole class of subtle bugs, for example, differences in byte order, field alignment, or encoding length.
+
+In practice, it means that if your function works in JavaScript, it will work in Compact:
```

Comment:
This is kind of meaningless.  The compiler-generated JS code is the same as the Compact contract.  There's no other meaning of "work in Compact" than (1) compile to JS and (2) run the JS.

Reply:
If we were to be precise about what it means for a function (circuit) to "work in Compact", it would have to be: it is compiled to JS and the JS code is run.

So it seems circular to say "if your function works in JS it will work in Compact".  That is just saying "if your function works in compiler-generated JS it will work in compiler-generated JS".